### PR TITLE
replica predictor scheduling plugin

### LINKED ISSUE #282: replica predictor scheduling plugin

### PR DESCRIPTION
**Pull request by yinsenyan**
#### What type of PR is this?
kind/feature

#### What this PR does / why we need it:
implement replica predictor scheduling plugin

#### Which issue(s) this PR fixes:

Fixes #282 
xref #344

### CODE REVIEW

File: pkg/scheduler/default_plugins.go [hunk 1]
```diff
@@ -32,8 +35,10 @@ func getDefaultPlugins() *schedulerapis.Plugins {
 		},
 		PostFilter:  schedulerapis.PluginSet{},
 		PreEstimate: schedulerapis.PluginSet{},
-		Estimate:    schedulerapis.PluginSet{},
-		PreScore:    schedulerapis.PluginSet{},
+		Estimate: schedulerapis.PluginSet{
+			Enabled: estimatorPlugin,
```
Comment (dixudx) [inline pkg/scheduler/default_plugins.go:39]: use inline struct instead.

File: pkg/scheduler/framework/plugins/defaultassigner/static.go [hunk 2]
```diff
@@ -49,7 +49,7 @@ func (pl *StaticAssigner) Name() string {
 func (pl *StaticAssigner) Assign(ctx context.Context, sub *appsapi.Subscription, finv *appsapi.FeedInventory, availableReplicas framework.TargetClusters) (framework.TargetClusters, *framework.Status) {
 	klog.V(3).InfoS("Attempting to assign replicas to clusters",
 		"subscription", klog.KObj(sub), "clusters", availableReplicas.BindingClusters)
-	if sub.Spec.DividingScheduling == nil || sub.Spec.DividingScheduling.Type != appsapi.StaticReplicaDividingType {
+	if sub.Spec.DividingScheduling == nil || (sub.Spec.DividingScheduling.Type != appsapi.StaticReplicaDividingType && sub.Spec.DividingScheduling.Type != appsapi.DynamicReplicaDividingType) {
```
Comment (dixudx) [inline pkg/scheduler/framework/plugins/defaultassigner/static.go:52]: This file is only for static scheduling. Please do not append dynamic scheduling here.

File: pkg/scheduler/framework/plugins/estimator/estimator.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+	clsapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
+	scheduler "github.com/clusternet/clusternet/pkg/apis/scheduler"
+	framework "github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces"
+	"github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/names"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Estimator is a plugin than checks if a subscripthon need resources
+type Estimator struct {
```
Comment (dixudx) [inline pkg/scheduler/framework/plugins/estimator/estimator.go:31]: Here should be added to `pkg/scheduler/framework/plugins/defaultassigner/dynamic.go.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/scheduler/framework/plugins/defaultassigner/assigner.go [hunk 4]
```diff
@@ -62,8 +62,14 @@ func (pl *StaticAssigner) Assign(ctx context.Context, sub *appsapi.Subscription,
 		result.BindingClusters = append(result.BindingClusters, name)
 	}
 
-	if err = StaticDivideReplicas(&result, sub, clusters, finv); err != nil {
-		return framework.TargetClusters{}, framework.AsStatus(err)
+	if sub.Spec.DividingScheduling.Type == appsapi.DynamicReplicaDividingType {
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/defaultassigner/assigner.go:65]: Better to change dynamic assigner to a new plugin as default dynamic assigner.

File: pkg/scheduler/framework/plugins/defaultassigner/algorithm.go [hunk 5]
```diff
@@ -17,6 +17,51 @@ type weightList struct {
...
+func dymamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) []int32 {
+	res := make([]int32, len(maxAvailableReplicas))
+	var sumAvailableReplicas int32 = sumArray(maxAvailableReplicas)
+
+	for k, _ := range res {
+		res[k] = (maxAvailableReplicas[k] / sumAvailableReplicas) * desiredReplicas
+	}
+
+	//The last cluster replenish replicas
+	res[len(res)-1] = desiredReplicas - sumArray(res[:len(res)-2])
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:53]: Better to even out the rest replicas among all candidate clusters.

File: pkg/scheduler/framework/plugins/defaultassigner/algorithm.go [hunk 6]
```diff
@@ -17,6 +17,51 @@ type weightList struct {
...
+	}
+	return nil
+}
+
+// dynamicDivideReplicas divides replicas by the MaxAvailableReplicas
+func dymamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) []int32 {
+	res := make([]int32, len(maxAvailableReplicas))
+	var sumAvailableReplicas int32 = sumArray(maxAvailableReplicas)
+
+	for k, _ := range res {
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:48]: `for k := range res`

File: pkg/scheduler/framework/plugins/defaultassigner/algorithm.go [hunk 7]
```diff
@@ -17,6 +17,51 @@ type weightList struct {
...
+			selected.Replicas[utils.GetFeedKey(order.Feed)] = []int32{}
+		}
+	}
+	return nil
+}
+
+// dynamicDivideReplicas divides replicas by the MaxAvailableReplicas
+func dymamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) []int32 {
+	res := make([]int32, len(maxAvailableReplicas))
+	var sumAvailableReplicas int32 = sumArray(maxAvailableReplicas)
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:46]: `sumAvailableReplicas := sumArray(maxAvailableReplicas)`

File: pkg/scheduler/framework/plugins/predictor/default_predictor.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+*/
+
+package predictor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/predictor/default_predictor.go:24]: Change the sequence for passing `go-fmt`

File: pkg/scheduler/framework/plugins/predictor/external_predictor.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	response, _ := strconv.Atoi(string(data))
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/predictor/external_predictor.go:64]: Should we handle errors?

File: pkg/scheduler/framework/plugins/predictor/predictor.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/names"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Predictor is a plugin than checks if a subscripthon need resources
+type Predictor struct {
+	handle framework.Handle
+}
+
+var _ framework.EstimatePlugin = &Predictor{}
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/predictor/predictor.go:35]: `PredictPlugin`

File: pkg/scheduler/framework/plugins/predictor/predictor.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (pl *Predictor) Name() string {
+	return names.Predictor
+}
+
+// PostFilter invoked post of the filter
+func (pl *Predictor) Estimate(ctx context.Context,
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/predictor/predictor.go:43]: `Predict`

**Comment by Garrybest** [comment]
Please rebase `main` branch and then fix check problems.

File: pkg/scheduler/framework/plugins/defaultassigner/algorithm.go [hunk 12]
```diff
@@ -17,6 +17,51 @@ type weightList struct {
...
+			selected.Replicas[utils.GetFeedKey(order.Feed)] = replicas
+		} else {
+			selected.Replicas[utils.GetFeedKey(order.Feed)] = []int32{}
+		}
+	}
+	return nil
+}
+
+// dynamicDivideReplicas divides replicas by the MaxAvailableReplicas
+func dymamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) []int32 {
```
Comment (Garrybest) [inline pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:44]: Seems like now this function only cares about disperse replicas in all candidate clusters. We need more strategies in the future so I have reserved the API before. It's OK, we may design this in next PR.

**Comment by silenceper** [comment]
Because the `maxAvailableReplicas` of the cluster is different each time the cluster is recalculated, there may be a situation:
For example, when the scheduler restarts, it may cause the number of workload replicas for each cluster to be recalculated and redeployed.

Should we avoid this from happening?

In addition, when the second update is performed, the maxAvailableReplicas of the computing cluster will be smaller than the actual value (if the replicas of the deployed workloads are not removed), resulting in inaccurate estimates of the number of replicas

**Comment by Garrybest** [comment]
> I have a question:
> 
> Because the `maxAvailableReplicas` of the cluster is different each time the cluster is recalculated, there may be a situation: For example, when the scheduler restarts, it may cause the number of workload replicas for each cluster to be recalculated and redeployed.
> 
> Should we avoid this from happening?
> 
> In addition, when the second update is performed, the maxAvailableReplicas of the computing cluster will be smaller than the actual value (if the replicas of the deployed workloads are not removed), resulting in inaccurate estimates of the number of replicas

You are right. We don't need to schedule these workloads which have been completely divided. So this condition should be skipped before the reconciler.

File: pkg/scheduler/framework/plugins/defaultassigner/algorithm.go (hunk 12 above)
Comment (yinsenyan) [reply pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:44]: If we have a new strategy, we can add a argument here and implement corresponding algorithm in next pr.

File: pkg/scheduler/framework/plugins/predictor/predictor.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+	clsapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
+	scheduler "github.com/clusternet/clusternet/pkg/apis/scheduler"
+	framework "github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces"
+	"github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/names"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Predictor is a plugin than checks if a subscripthon need resources
```
Comment (dixudx) [inline pkg/scheduler/framework/plugins/predictor/predictor.go:29]: typo subscripthon

Please rephrase the comment.

File: pkg/scheduler/framework/plugins/predictor/predictor.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+var _ framework.PredictPlugin = &Predictor{}
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (pl *Predictor) Name() string {
+	return names.Predictor
+}
+
+// PostFilter invoked post of the filter
```
Comment (dixudx) [inline pkg/scheduler/framework/plugins/predictor/predictor.go:41]: mismatching caller name and comment.

**Review by dixudx — APPROVED** [review]
Thanks for the work!

**Pull request merged**